python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -701,6 +702,45 @@ def cummax(self):
         """Get the column-wise cumulative maximum value in each group."""
         return self.agg("cummax")
 
+    def shift(self, periods=1, freq=None, axis=0, fill_value=None):
+        """
+        Shift each group by ``periods`` positions.
+
+        Parameters
+        ----------
+        periods : int, default 1
+            Number of periods to shift.
+        freq : str, unsupported
+        axis : 0, axis to shift
+            Shift direction. Only row-wise shift is supported
+        fill_value : optional
+            The scalar value to use for newly introduced missing values. Should
+            match the dtype of columns to fill.
+
+        Returns
+        -------
+        Series or DataFrame
+            Object shifted within each group.
+
+        Notes
+        -----
+        Parameter ``freq`` is unsupported.
+        """
+
+        if freq is not None:
+            raise NotImplementedError("Parameter freq is unsupported.")
+
+        if not axis == 0:
+            raise NotImplementedError("Only axis=0 is supported.")
+
+        data_names = [
+            x for x in self.obj._column_names if x not in self.grouping.names
+        ]
+        data = self.obj._data.select_by_label(data_names)
+        fill_values = [fill_value] * len(data)
```

Comment:
`len(data)` is pretty ambiguous here. Is it the number of columns or number of rows of data?

Should we be generating `fill_values` on the GPU instead?

Reply:
That would be the number of columns here. One `fill_value` for a column. Not sure what generating `fill_values` on GPU mean? `as_device_scalar` is used later in Cython to make sure the data is present on device side.

Reply:
I could be wildly incorrect, as I'm not familiar with the groupby apis, but this name _seems_ more informative... ?

`columns = self.obj._data.select_by_label(...)` ?

"`data`" makes it sound like the individual elements are being assigned values, hence the question about doing it on the GPU, I presume.

Reply:
How about `value_columns`? The primary thing I'm trying to separate out here is there are `key` columns and `value` columns. The `shift` operation should only operate on value columns.